pandas/core/window/numba_.py

```diff
@@ -0,0 +1,127 @@
+import types
+from typing import Any, Callable, Dict, Optional, Tuple
+
+import numpy as np
+
+from pandas._typing import Scalar
+from pandas.compat._optional import import_optional_dependency
+
+
+def make_rolling_apply(
+    func: Callable[..., Scalar],
+    args: Tuple,
+    nogil: bool,
+    parallel: bool,
+    nopython: bool,
+):
+    """
+    Creates a JITted rolling apply function with a JITted version of
+    the user's function.
+
+    Parameters
+    ----------
+    func : function
+        function to be applied to each window and will be JITed
+    args : tuple
+        *args to be passed into the function
+    nogil : bool
+        nogil parameter from engine_kwargs for numba.jit
+    parallel : bool
+        parallel parameter from engine_kwargs for numba.jit
+    nopython : bool
+        nopython parameter from engine_kwargs for numba.jit
+
+    Returns
+    -------
+    Numba function
+    """
+    numba = import_optional_dependency("numba")
+
+    if parallel:
+        loop_range = numba.prange
+    else:
+        loop_range = range
+
+    if isinstance(func, numba.targets.registry.CPUDispatcher):
+        # Don't jit a user passed jitted function
+        numba_func = func
+    else:
+
+        @numba.generated_jit(nopython=nopython, nogil=nogil, parallel=parallel)
```

Comment:
@stuartarchibald sorry for the ping, but I see that `generated_jit` has been deprecated in numba 0.57. IIRC you helped me add this a while back and am lost on how to write this in terms of `overload`

Reply:
@mroeschke no problem, I can try and help with this. I think it needs to look a bit like this (for reference, this is untested, I am just guessing from the context! Also, the `pandas` variant is obviously wrapped to close over some configuration which I've omitted, so consider this as the function body of `make_rolling_apply`. I've left comments inline to try and explain what's going on):

```python
import types
import numpy as np
from numba.extending import overload, is_jitted
from numba import njit
import numba


# this provides a local definition to overload
def overload_target(window, *_args):
    # If JIT is disabled, this function will run, so write the implementation here!
    pass


nopython = True
nogil = True
parallel = False

# pretend this is an arg to `make_rolling_apply`
def func(window, *args):
    return window * 2 + args[0]


@overload(overload_target, jit_options={'nopython':nopython, 'nogil':nogil,
                                        'parallel':parallel})
def ol_overload_target(window, *_args):
    # This function "overloads" `overload_target`, whenever the Numba compiler
    # "sees" `overload_target` it will use this function.

    # Using `is_jitted` to avoid `isinstance` on
    # `numba.targets.registry.CPUDispatcher` as that may be considered an
    # internal Numba detail.
    if is_jitted(func):
        # it's already JIT compiled so just reference it
        overload_target_impl = func
    elif getattr(np, func.__name__, False) is func or isinstance(
        func, types.BuiltinFunctionType
    ):
        # it's a NumPy function or builtin so just reference it
        overload_target_impl = func
    else:
        # it's a Python function, so register it as JIT compilable and reference
        # that
        overload_target_impl = numba.jit(func, nopython=nopython, nogil=nogil)

    # This is the Numba implementation of the overload, it will just be JIT
    # compiled whenever the compiler "sees" a reference to "overload_target" in
    # code it is compiling.
    def impl(window, *_args):
        return overload_target_impl(window, *_args)

    return impl


# demo

@njit
def roll_apply(window, *_args):
    return overload_target(window, *_args)


print(roll_apply(np.arange(10.), 1.23))
```

`@overload` is basically saying to Numba "when you see this specific python function (the one in the first argument in the `@overload` decorator) use this implementation". The concept about there being a "typing" part that can be used to dispatch different variants based on type is exactly the same as in `@generated_jit`. The largest difference is what happens if the JIT compiler is turned off. In the case of `@overload` the python function being overloaded will run, i.e., the code just executes as would be expected in the interpreter. Whereas in the case of `@generated_jit`, because the pure python implementation and the Numba implementation are the same function, if you turn the JIT compiler off it will just break (the value returned when calling a `@generated_jit` function is a function implementing the Numba specialisation). Essentially, `@generated_jit` is like doing `@overload` but the function being decorated is also the function being overloaded.

Hope this helps?

Reply:
Thanks for the reply! We had a PR recently that refactored this to use `extending.register_jittable`. Would that be a sufficient alternative? https://github.com/pandas-dev/pandas/pull/53455/files

Reply:
No problem! I just took a look at the patch above, I think it'd work but think it might lose some of the dispatch ability offered by `generated_jit`/`overload`. As I understand it, the original code would have let a NumPy function or a built-in be passed in as the "user function", whereas I think the `register_jittable` version requires a user defined Python function. It may be that the `register_jittable` version is a sufficient alternative for the need/use cases in practice, in which case, it seems appropriate.

Reply:
Great thanks for the context! Yeah this function should expect a custom UDF so thanks for the confirmation

Reply:
Glad to get this resolved, thanks for confirming too! It sounds like the replacement above is appropriate. If there are any more issues/queries feel free to open issues on the Numba issue tracker (or ping here!).

Reply:
@stuartarchibald I'm running into a [rolling apply issue](https://github.com/pandas-dev/pandas/issues/55247) with pandas 2.1.1 and numba 0.58 that might be related. Discussion is here:
https://numba.discourse.group/t/pandas-source-of-old-style-error-capturing-warning/2169/8